### PR TITLE
[backend] container upload more flexible

### PR DESCRIPTION
**Pull request by M0ses**
With this patch you can configure the container upload with skopeo
more flexible.

* You can define regex for the project names to be uploadeded
* You can define multiple registries to upload to.

### CODE REVIEW

File: src/backend/BSConfig.pm.template [hunk 1]
```diff
@@ -238,4 +238,36 @@ our $servicedispatch = 1;
 #   repository_base => '',
 # };
 
+# our $container_registries = {
```
Comment (ChrisBr) [inline src/backend/BSConfig.pm.template]: Why are we not going to store that in the frontend?

We could even make this dynamic and store it in the database, no?

**Comment by hennevogel** [comment]
I'm really starting to worry about this feature being completely developed inside the backend. Why do I have to login to the OBS server and edit backend config files regexe's to be able to do this? Why can't I do this on `/configuration` and with an attribute?

**Comment by adrianschroeter** [comment]
@hennevogel it is the same as with the stage server, also a pure backend thing.
It would be even a security issue if you can change this via the api.

**Comment by hennevogel** [comment]
@adrianschroeter I understand that it's also about publishing and that so far we have done publishing from the backend. But why do we have to keep doing this? Especially if it involves configuring access to a service and project names...

And why on earth would this be a security issue if an authorized admin can change this through the API?

**Comment by adrianschroeter** [comment]
On Mittwoch, 18. Oktober 2017, 13:21:10 CEST wrote Henne Vogelsang:
> @adrianschroeter I understand that it's also about publishing and that so far we have done publishing from the backend. But why do we have to keep doing this? Especially if it involves configuring access to a service and project names...

For pure bandwidth (avoiding multiple transfers and container layer filtering), 
network setup (only backends reach stage servers), storage, 
scalability(via multiple backends), reliablity, .... reasons.

> And why on earth would this be a security issue if an authorized admin can change this through the API?

Because it is only an OBS admin user and not a backend server host root user who has permission to change 
backend network related configs.

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by M0ses** [comment]
@mlschroe Please review again

**Comment by hennevogel** [comment]
> For pure bandwidth (avoiding multiple transfers and container layer filtering), network setup (only backends reach stage servers), storage, scalability(via multiple backends), reliablity, .... reasons.

What has that to do with where we configure this and expose this configuration? The backend is perfectly capable of reading things from `/configuration` right?

> Because it is only an OBS admin user and not a backend server host root user who has permission to change backend network related configs.

In my experience this distinction is artificial and they are always the same people. And what is a `backend  network`?

The reason this makes me nervous is that we are configuring a service and have no way to expose this configuration in the future to our users somehow, short of writing an interface to BSConfig.pm. It would be a good point in time to stop doing this and move forward in a unified fashion…

**Comment by adrianschroeter** [comment]
On Mittwoch, 18. Oktober 2017, 14:01:09 CEST wrote Henne Vogelsang:
> > For pure bandwidth (avoiding multiple transfers and container layer filtering), network setup (only backends reach stage servers), storage, scalability(via multiple backends), reliablity, .... reasons.
> 
> What has that to do with where we configure this and expose this configuration? The backend is perfectly capable of reading things from `/configuration` right?

yes, but it keeps it's own configs beside for things which never should get changed from anyone else, but the root user.

note that this host configured here might not even reachable for the ruby code instance.

> > Because it is only an OBS admin user and not a backend server host root user who has permission to change backend network related configs.
> 
> In my experience this distinction is artificial and they are always the same people. And what is a `backend  network`?

You lack some basic understand of security mechanics here then.... the backend network is the (optional seperated) 
network where backend components speak to each other.

> The reason this makes me nervous is that we are configuring a service and have no way to expose this configuration in the future to our users somehow, short of writing an interface to BSConfig.pm. It would be a good point in time to stop doing this and move forward in a unified fashion…

Users will never ever be able to tell the publisher where to push content.
If you want this, it must happen in another instance and it must use a registry
proxy instance. But this entirely out of scope here.

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by hennevogel** [comment]
> yes, but it keeps it's own configs beside for things which never should get changed from anyone else, but the root user.

Why? What is the point?

> note that this host configured here might not even reachable for the ruby code instance.

Again, this has nothing to do with this.

**Comment by adrianschroeter** [comment]
On Mittwoch, 18. Oktober 2017, 14:44:37 CEST wrote Henne Vogelsang:
> > yes, but it keeps it's own configs beside for things which never should get changed from anyone else, but the root user.
> 
> Why? What is the point?

Just as an example, would you think it would be wise to design an api
interface for the OBS Admin user which causes changes of the apache 
vhost configuration?

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by hennevogel** [comment]
> you think it would be wise to design an api interface for the OBS Admin user which causes changes of the apache vhost configuration?

Of course not. First this is something that rarely changes and requires a restart of apache and apache is not part of our application, it's just the application server we use to serve the rails app. I can also just skip it and run the RoR app without or with some other application service...

**But** the backend and hence the publisher are part of the OBS application and changing where the publisher would push things, including based on which conditions, doesn't require a restart and is something that I would imagine will change quite often and even could be exposed to users via an easier interface that a perl file to edit...

**Comment by hennevogel** [comment]
I would also really like if you could answer the question I have asked:

1. Why do you think we should keep publishing a backend only feature?
2. Why do you think it is a security risk if an authorized admin user can change publisher configuration?
3. Why will users never ever be able to tell the publisher where to push content?

**Comment by adrianschroeter** [comment]
On Mittwoch, 18. Oktober 2017, 15:09:17 CEST wrote Henne Vogelsang:
> > you think it would be wise to design an api interface for the OBS Admin user which causes changes of the apache vhost configuration?
> 
> Of course not. First this is something that rarely changes and requires a restart of apache and apache is not part of our application, it's just the application server we use to serve the rails app. I can also just skip it and run the RoR app without or with some other application service...
> 
> **But** the backend and hence the publisher are part of the OBS application and changing where the publisher would push things, including based on which conditions, doesn't require a restart and is something that I would imagine will change quite often and even could be exposed to users via an easier interface that a perl file to edit...

A change would mean that we would also setup a new registry server. So, no, I don't think
we will change it that often.

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by hennevogel** [comment]
> A change would mean that we would also setup a new registry server. So, no, I don't think we will change it that often.

This might be true for your single use case for your instance, I'm talking about this in general.

**Comment by hennevogel** [comment]
Oh well, I give up trying to work with you guys...

**Comment by adrianschroeter** [comment]
On Mittwoch, 18. Oktober 2017, 15:18:22 CEST wrote Henne Vogelsang:
> I would also really like if you could answer the question I have asked:
> 

I could write really long text here, but trying to stay short here and
give just one reason each out of many...

> 1. Why do you think we should keep publishing a backend only feature?

for one because the api will not write back binary state results (like publish success failure)
to the backend, neither make use the possible configured hooks there.

> 2. Why do you think it is a security risk if an authorized admin user can change publisher configuration?

because you could attack other parts in the internal backend network which
is shielded. being an OBS admin need to mean that you can trace it's
actions in the backend.

> 3. Why will users never ever be able to tell the publisher where to push content?

Because the publisher (or better the IO of the publisher) will always stay a 
bottle neck. the stage servers must remain to have a guaranteed answer time to
avoid DoS attacks (which could happen easily even when not intended).

If you want to upload to more places, you would need to do this from the
stage servers. And that would not be a registry specific thing, we have this
already with the repo pusher. However, it makes most likely little sense
in docker/registry world, since you would just setup a registry proxy 
for caching here.

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Pull request merged**